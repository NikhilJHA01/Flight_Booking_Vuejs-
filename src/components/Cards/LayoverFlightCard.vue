<template>
  <section class="layoverFlightCard">
    <el-card class="card" shadow="always">
      <!-- Display Overview of Layover Journey -->
      <div class="card__body">
        <div class="flex-column">
          <svg class="icon icon-aircraft-take-off">
            <use xlink:href="/sprite.svg#icon-aircraft-take-off"></use>
          </svg>
        </div>
        <div class="flex-column">
          <span class="primary--text">Multiple</span
          ><span class="showDetailsLink" @click="toggleDetails()">{{
            showDetails ? "Hide Details" : "Show Details"
          }}</span>
        </div>
        <div class="flex-column">
          <span class="primary--text">{{
            layOverFlights[0].departureTime
          }}</span
          ><span class="secondary--text">{{
            layOverFlights[0].origin | removeCityCode
          }}</span>
        </div>
        <div class="flex-column">
          <span class="primary--text">{{ layOverFlights[1].arrivalTime }}</span
          ><span class="secondary--text">{{
            layOverFlights[1].destination | removeCityCode
          }}</span>
        </div>
        <div class="flex-column">
          <span class="primary--text">{{ totalHoursOfJourney }}</span
          ><span class="secondary--text">Total Duration</span>
        </div>
        <div class="flex-column">
          <span class="primary--text red-color">
            &#x20b9;{{ totalPriceOfTickets }}</span
          >
          <span class="secondary--text">{{ ticketCount }} Ticket</span>
        </div>
        <el-button
          type="success"
          @click="bookingSuccess()"
          class="defaultBookButton"
          >Book</el-button
        >
      </div>

      <!-- Display Multiple flights when clicked on show details -->
      <div v-if="showDetails" class="layoverFlightCard__details">
        <el-card
          class="card layoverFlightCard__layoutCard"
          v-for="flight in layOverFlights"
          :key="flight.flightNo"
        >
          <div class="card__body ">
            <div class="flex-column">
              <svg class="icon icon-aircraft-take-off">
                <use xlink:href="/sprite.svg#icon-aircraft-take-off"></use>
              </svg>
            </div>
            <div class="flex-column">
              <span class="primary--text">{{ flight.name }}</span
              ><span class="secondary--text">{{ flight.flightNo }}</span>
            </div>
            <div class="flex-column">
              <span class="primary--text">{{ flight.departureTime }}</span
              ><span class="secondary--text">{{
                flight.origin | removeCityCode
              }}</span>
            </div>
            <div class="flex-column">
              <span class="primary--text">{{ flight.arrivalTime }}</span
              ><span class="secondary--text">{{
                flight.destination | removeCityCode
              }}</span>
            </div>
            <div class="flex-column">
              <span class="primary--text">{{ flight.totalHours }}</span
              ><span class="secondary--text">Non Stop</span>
            </div>
          </div>
        </el-card>
        <div class="layoverFlightCard--layoverTime">
          Layover Time {{ layOverTime }}
        </div>
      </div>
    </el-card>
  </section>
</template>

<script>
import moment from "moment";
import { MomentService } from "../../utility/momentUtil";

export default {
  name: "LayoverFlightCard",
  data() {
    return {
      showDetails: false
    };
  },
  props: {
    layOverFlights: {
      type: Array,
      required: true
    },
    ticketCount: {
      type: Number,
      required: true
    }
  },
  computed: {
    // Calculate time between first flights arrival and second's departure
    layOverTime() {
      return MomentService.convertHoursFormat(
        this.layOverFlights[0].arrivalTime,
        this.layOverFlights[1].departureTime
      );
    },

    // Calculate time from start of journey to End when taking multiple flights
    totalHoursOfJourney() {
      return MomentService.convertHoursFormat(
        this.layOverFlights[0].departureTime,
        this.layOverFlights[1].arrivalTime
      );
    },

    // Calculate Sum of prices of multiple flights taken in journey * total travellers
    totalPriceOfTickets() {
      return (
        this.layOverFlights.reduce((accumulator, currentValue) => {
          return accumulator + currentValue.price;
        }, 0) * this.ticketCount
      );
    }
  },
  methods: {
    // Emit event to Parent To Open the Booking dialog Message
    bookingSuccess() {
      this.$emit("showBookingMessage");
    },
    // Toggle Show Details and Hide details
    toggleDetails() {
      this.showDetails = !this.showDetails;
    }
  }
};
</script>

<style lang="scss" scoped>
.layoverFlightCard {
  .icon {
    @extend %default-icon;
    fill: $primary-main;
  }
  .card {
    margin: 10px;
    &__body {
      @extend %default-card-body;
      margin-bottom: 10px;
    }
  }
  &__layoutCard {
    margin-bottom: 2px;
    background-color: #ebeef5;
    box-shadow: none;
  }
  &--layoverTime {
    left: 50%;
    position: absolute;
    transform: translate(-50%, -50%);
    top: 50%;
    padding: 8px;
    border-radius: 5px;
    color: #606266;
    font-size: 13px;
    background-color: white;
  }
  &__details {
    position: relative;
  }
  .showDetailsLink {
    cursor: pointer;
    color: $primary-main;
    font-size: 12px;
    padding: 2px;
    @include phone() {
      font-size: 9px;
    }
  }
}
</style>
