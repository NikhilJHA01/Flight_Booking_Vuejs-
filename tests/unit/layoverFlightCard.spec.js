import { mount } from "@vue/test-utils";
import LayoverFlightCard from "@/components/Cards/LayoverFlightCard";
import { layOverFlightdata } from "./data/flightTestData";

describe("LayoverFlightCard", () => {
  const wrapper = mount(LayoverFlightCard, {
    propsData: {
      layOverFlights: layOverFlightdata,
      ticketCount: 1
    }
  });
  it("renders the Non Stop Card", async () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("Fires a Book event on succesful book from Layover Flights", async () => {
    wrapper.find("el-button").trigger("click");
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted("showBookingMessage")).toBeTruthy();
  });

  it("check NonStop Flights card has been rendered and it contains svg", () => {
    const form = wrapper.find("svg");
    expect(form.exists()).toBe(true);
  });

  it("check NonStop Flights card has been rendered and it contains showLink", () => {
    const form = wrapper.find(".showDetailsLink");
    expect(form.exists()).toBe(true);
  });

  it("Click on show details link", async () => {
    expect(wrapper.vm.showDetails).toBe(false);
    wrapper.find(".showDetailsLink").trigger("click");
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.showDetails).toBe(true);
    expect(wrapper.emitted("showBookingMessage")).toBeTruthy();
  });
});
