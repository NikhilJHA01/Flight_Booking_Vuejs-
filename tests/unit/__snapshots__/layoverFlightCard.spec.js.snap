// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayoverFlightCard renders the Non Stop Card 1`] = `
<section class="layoverFlightCard">
  <el-card shadow="always" class="card">
    <div class="card__body">
      <div class="flex-column"><svg class="icon icon-aircraft-take-off">
          <use xlink:href="/sprite.svg#icon-aircraft-take-off"></use>
        </svg></div>
      <div class="flex-column"><span class="primary--text">Multiple</span><span class="showDetailsLink">Show Details</span></div>
      <div class="flex-column"><span class="primary--text">8:15</span><span class="secondary--text">Pune (PNQ)</span></div>
      <div class="flex-column"><span class="primary--text">9:15</span><span class="secondary--text">Bengaluru (BLR)</span></div>
      <div class="flex-column"><span class="primary--text">1 Hr 0 min</span><span class="secondary--text">Total Duration</span></div>
      <div class="flex-column"><span class="primary--text red-color">
          â‚¹7098</span> <span class="secondary--text">1 Ticket</span></div>
      <el-button type="success" round="">Book</el-button>
    </div>
    <!---->
  </el-card>
</section>
`;
